src/app/clusters/door-lock-server/door-lock-server.cpp

```diff
@@ -140,30 +145,107 @@ bool emberAfDoorLockClusterLockDoorCallback(chip::app::CommandHandler * commandO
                                             const chip::app::ConcreteCommandPath & commandPath,
                                             const chip::app::Clusters::DoorLock::Commands::LockDoor::DecodableType & commandData)
 {
-    emberAfDoorLockClusterPrintln("Received Lock Door command (not implemented)");
+    emberAfDoorLockClusterPrintln("Received Lock Door command");
+    bool success = false;
 
-    // TODO: Implement door locking by calling emberAfPluginDoorLockOnDoorLockCommand
+    chip::EndpointId endpoint = commandPath.mEndpointId;
 
-    emberAfSendImmediateDefaultResponse(EMBER_ZCL_STATUS_SUCCESS);
-    return true;
+    bool require_pin = false;
+    Attributes::RequirePINforRemoteOperation::Get(endpoint, &require_pin);
+
+    if (commandData.pinCode.HasValue())
+    {
+        // TODO: Search through list of stored PINs and check each.
+        if (mPin.data_equal(commandData.pinCode.Value()))
+        {
+            success = emberAfPluginDoorLockOnDoorLockCommand(endpoint, commandData.pinCode);
+        }
+        else
+        {
+            success = false; // Just to be explicit. success == false at this point anyway
+        }
+    }
+    else
+    {
+        if (!require_pin)
+        {
+            success = emberAfPluginDoorLockOnDoorLockCommand(endpoint, commandData.pinCode);
+        }
+        else
+        {
+            success = false;
+        }
+    }
+
+    emberAfSendImmediateDefaultResponse(success ? EMBER_ZCL_STATUS_SUCCESS : EMBER_ZCL_STATUS_FAILURE);
+
+    // Should be called after reponse sent (indicating command complete) since
+    // the attribute Set call in SetLockState will error out if the command is
+    // still pending
+    if (success)
+        return DoorLockServer::Instance().SetLockState(endpoint, DlLockState::kLocked) == EMBER_ZCL_STATUS_SUCCESS;
+    return success;
```

Comment:
Wait, this is not right.  If success is false, we already sent a response above, but returning false here will send one _again_.

A false return only happens when we could not process things enough to even send a failure response, basically.  This should be "return true".  And the SetLockState() call should probably not be returning, if it's after the default response is sent.

Reply:
Ok. Reworked to send default response after SetLockState is called (and return checked) and just have the callback return true.